### PR TITLE
lp1766042: Fix FLAC decoding (-6 dB) and upgrade DB schema

### PR DESCRIPTION
**Pull request by uklotzde**
- https://bugs.launchpad.net/mixxx/+bug/1766042
- Follow-up of #1633
- Database schema upgrade that resets the replay gain info of all FLAC files

### CODE REVIEW

**Comment by uklotzde** [comment]
The case-insensitive comparison of the file type is just for additional safety. Mixxx already converts all file extensions into lower-case during while importing files.

**Comment by Pegasus-RPG** [comment]
What about FLAC files that had their ReplayGain added outside of Mixxx (e.g. with metaflac?) Is there a way to tell so not all have to be invalidated?

**Comment by Pegasus-RPG** [comment]
Otherwise LGTM.

**Comment by uklotzde** [comment]
The database upgrade does neither reset exported replay gain info in file tags nor the waveforms (as mentioned by Jamie). We should add a release note to ask for resetting/clearing and re-analyzing both manually.

**Comment by uklotzde** [comment]
Ups, sorry, my last comment mentioned another aspect.

Unfortunately we don't know the actual source of the replay gain info in our database. A warning in the release notes accompanied by guidelines how to compensate for this will be required. Depending on their use case users should be asked to either re-import (which cannot be limited to replay gain) or re-analyze (see my previous comment). Re-analyzing with the new EBU R128 algorithm is the preferred solution.

**Comment by Pegasus-RPG** [comment]
It would be good to also mention in that note to double-check the setting of the "Re-analyze and override an existing value" checkbox in the prefs.

**Comment by uklotzde** [comment]
As expected the replay gain is lowered by 6 dB and the actual values are divided by 2.

**Comment by uklotzde** [comment]
But I discovered another bug especially for this use case while testing: Replay gain is only exported/written into file tags if it is valid! Otherwise the information in the file tags is left untouched instead of deleting it as expected. Clearing the ReplayGain in Mixxx does not propagate into the files when metadata export is enabled.

I will file a bug shortly!

File: src/sources/soundsourceflac.cpp [hunk 1]
```diff
@@ -435,9 +435,18 @@ void SoundSourceFLAC::flacMetadata(const FLAC__StreamMetadata* metadata) {
...
         if (kBitsPerSampleDefault == m_bitsPerSample) {
             // not set before
-            m_bitsPerSample = bitsPerSample;
-            m_sampleScaleFactor = CSAMPLE_PEAK
-                    / CSAMPLE(FLAC__int32(1) << bitsPerSample);
+            if (bitsPerSample > 0) {
+                m_bitsPerSample = bitsPerSample;
+                // Range of signed(!) sample values: [2 ^ (bitsPerSample - 1), 2 ^ (bitsPerSample - 1) - 1]
+                // See also: https://bugs.launchpad.net/mixxx/+bug/1766042
+                const auto maxAbsSampleValue = FLAC__int32(1) << (bitsPerSample - 1);
```
Comment (daschuer) [inline src/sources/soundsourceflac.cpp]: The cast here and below are IMHO unnecessary. If you think they are, use static_cast.

**Comment by daschuer** [comment]
Only a minor readability issue. I would just ignore the waveform issue. Because only the waveform gain is effected. I hint in the readme should be sufficient.

**Comment by daschuer** [comment]
Al ci fails, because of a timeout during dB schema upgrade.

**Comment by uklotzde** [comment]
Looks like I have to adjust the UPDATE statement:
"Beginning in SQLite version 3.15.0 (2016-10-14), an assignment in the SET clause can be a parenthesized list of column names on the left and a row value of the same size on the right."

**Comment by daschuer** [comment]
LGTM Is this ready for merge?

**Comment by uklotzde** [comment]
Considered complete and finished.

**Pull request merged**